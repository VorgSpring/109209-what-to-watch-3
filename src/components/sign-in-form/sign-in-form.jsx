import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {withForm} from '../../hoc/with-form/with-form';
import {FormFields} from '../../constants/form-fields';
import {sendAuthorizationData} from '../../operations/authorization/authorization';
import {
  getErrorAuthorizationSelector,
  getLoadingStatusAuthorizationSelector
} from '../../selectors/authorization/authorization';

export class SingInFormComponent extends PureComponent {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(e) {
    e.preventDefault();

    const {
      data,
      validate,
      authorization
    } = this.props;

    const isValid = validate();

    if (isValid) {
      authorization(data);
    }
  }

  render() {
    const {onChange, data, errors, responseError} = this.props;
    const {email, password} = data;
    const {email: emailError, password: passwordError} = errors;

    return (
      <form
        action='#'
        className='sign-in__form'
        onChange={onChange}
        onSubmit={this.handleSubmit}
      >
        <div className='sign-in__message'>
          {emailError && <p>{emailError}</p>}
          {passwordError && <p>{passwordError}</p>}
          {responseError && <p>{responseError}</p>}
        </div>
        <div className='sign-in__fields'>
          <div className={`sign-in__field ${emailError && `sign-in__field--error`}`}>
            <input
              className='sign-in__input'
              type={FormFields.EMAIL}
              placeholder='Email address'
              name={FormFields.EMAIL}
              id={FormFields.EMAIL}
              defaultValue={email}
            />
            <label
              className='sign-in__label visually-hidden'
              htmlFor={FormFields.EMAIL}
            >
              Email address
            </label>
          </div>
          <div className={`sign-in__field ${passwordError && `sign-in__field--error`}`}>
            <input
              className='sign-in__input'
              type={FormFields.PASSWORD}
              placeholder='Password'
              name={FormFields.PASSWORD}
              id={FormFields.PASSWORD}
              defaultValue={password}
            />
            <label
              className='sign-in__label visually-hidden'
              htmlFor={FormFields.PASSWORD}
            >
              Password
            </label>
          </div>
        </div>
        <div className='sign-in__submit'>
          <button
            className='sign-in__btn'
            type='submit'
          >
            Sign in
          </button>
        </div>
      </form>
    );
  }
}

SingInFormComponent.propTypes = {
  data: PropTypes.shape({
    email: PropTypes.string.isRequired,
    password: PropTypes.string.isRequired,
  }).isRequired,
  errors: PropTypes.shape({
    email: PropTypes.string,
    password: PropTypes.string
  }).isRequired,
  responseError: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  authorization: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  validate: PropTypes.func.isRequired,
};

const initialFields = [FormFields.EMAIL, FormFields.PASSWORD];
export const SingInFormWrapper = withForm(SingInFormComponent, initialFields);

const mapStateToProps = (state) => ({
  responseError: getErrorAuthorizationSelector(state),
  isLoading: getLoadingStatusAuthorizationSelector(state)
});

const mapDispatchToProps = (dispatch) => ({
  authorization: (data) => dispatch(sendAuthorizationData(data))
});

export const SingInForm = connect(
    mapStateToProps, mapDispatchToProps
)(SingInFormWrapper);
