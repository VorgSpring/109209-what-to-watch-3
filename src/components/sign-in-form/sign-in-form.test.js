import React from 'react';
import {SingInForm} from './sign-in-form.jsx';
import renderer from 'react-test-renderer';

describe(`SingInForm`, () => {
  const data = {
    email: `blah`,
    password: `blah`
  };
  const errors = {
    email: null,
    password: null
  };
  const responseError = null;
  const isLoading = false;
  const authorization = () => {};
  const onChange = () => {};
  const validate = () => {};

  it(`should renders correctly`, () => {
    const tree = renderer.create(
        <SingInForm
          data={data}
          errors={errors}
          responseError={responseError}
          isLoading={isLoading}
          authorization={authorization}
          onChange={onChange}
          validate={validate}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should renders correctly with errors`, () => {
    const notEmptyErrors = {
      email: `blah`,
      password: `blah`
    };

    const tree = renderer.create(
        <SingInForm
          data={data}
          errors={notEmptyErrors}
          responseError={responseError}
          isLoading={isLoading}
          authorization={authorization}
          onChange={onChange}
          validate={validate}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
