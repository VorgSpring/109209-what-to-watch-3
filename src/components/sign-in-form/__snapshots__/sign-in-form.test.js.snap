// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingInForm should renders correctly 1`] = `
<form
  action="#"
  className="sign-in__form"
  onChange={[Function]}
  onSubmit={[Function]}
>
  <div
    className="sign-in__message"
  />
  <div
    className="sign-in__fields"
  >
    <div
      className="sign-in__field null"
    >
      <input
        className="sign-in__input"
        defaultValue="blah"
        id="email"
        name="email"
        placeholder="Email address"
        type="email"
      />
      <label
        className="sign-in__label visually-hidden"
        htmlFor="email"
      >
        Email address
      </label>
    </div>
    <div
      className="sign-in__field null"
    >
      <input
        className="sign-in__input"
        defaultValue="blah"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <label
        className="sign-in__label visually-hidden"
        htmlFor="password"
      >
        Password
      </label>
    </div>
  </div>
  <div
    className="sign-in__submit"
  >
    <button
      className="sign-in__btn"
      type="submit"
    >
      Sign in
    </button>
  </div>
</form>
`;

exports[`SingInForm should renders correctly with errors 1`] = `
<form
  action="#"
  className="sign-in__form"
  onChange={[Function]}
  onSubmit={[Function]}
>
  <div
    className="sign-in__message"
  >
    <p>
      blah
    </p>
    <p>
      blah
    </p>
    <p>
      blah-blah
    </p>
  </div>
  <div
    className="sign-in__fields"
  >
    <div
      className="sign-in__field sign-in__field--error"
    >
      <input
        className="sign-in__input"
        defaultValue="blah"
        id="email"
        name="email"
        placeholder="Email address"
        type="email"
      />
      <label
        className="sign-in__label visually-hidden"
        htmlFor="email"
      >
        Email address
      </label>
    </div>
    <div
      className="sign-in__field sign-in__field--error"
    >
      <input
        className="sign-in__input"
        defaultValue="blah"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <label
        className="sign-in__label visually-hidden"
        htmlFor="password"
      >
        Password
      </label>
    </div>
  </div>
  <div
    className="sign-in__submit"
  >
    <button
      className="sign-in__btn"
      type="submit"
    >
      Sign in
    </button>
  </div>
</form>
`;
